src/librustc_lint/types.rs

```diff
@@ -340,6 +421,69 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TypeLimits {
                 _ => false,
             }
         }
+
+        fn get_bin_hex_repr(cx: &LateContext, lit: &ast::Lit) -> Option<String> {
+            if let Some(src) = cx.sess().codemap().span_to_snippet(lit.span).ok() {
+                if let Some(firstch) = src.chars().next() {
+                    if let Some(0) = char::to_digit(firstch, 10) {
+                        if let Some(base) = src.chars().nth(1) {
+                            if base == 'x' || base == 'b' {
+                                return Some(src);
+                            }
+                        }
+                    }
+                }
+            }
+
+            None
+        }
+
+        fn get_fitting_type<'a>(
+            t: &ty::TypeVariants,
+            val: u128,
+            negative: bool,
+        ) -> Option<ty::TypeVariants<'a>> {
+            use syntax::ast::IntTy::*;
+            use syntax::ast::UintTy::*;
+            macro_rules! find_fit {
+                ($ty:expr, $val:expr, $negative:expr,
+                 $($type:ident => [$($utypes:expr),*] => [$($itypes:expr),*]),+) => {
+                    {
+                        let _neg = if negative { 1 } else { 0 };
```

Comment:
nit: preceding underscores suppress dead code warnings, so I'd avoid them if that is not intended.

Also the PR has a succint description for what this function does, which you could copy into a comment.

Reply:
The underscore is intended: for finding a fitting unsigned type, we don't look at integer types so the expanded macro never uses `_neg` in this case. So the line 
```rust
if let &ty::TyUint(u) = t {
    return find_fit!(u, val, negative
...}
```
gives a warning. If there is a better way to do this, it would be nice to know :)

And yeah maybe I should have wrote a few more comments. Will do this, while implementing the other suggestions